sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1847,3 +1848,25 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
       empty(ll)
   }
 }
+
+/**
+ * Generates filters for exploded expression, such that rows that would have been removed
+ * by this [[Generate]] can be removed earlier - before joins and in data sources.
+ */
+object InferFiltersFromGenerate extends Rule[LogicalPlan] {
```

Comment:
Why did you create a new rule instead of improving the exisint one, `InferFiltersFromConstraints`?

Reply:
The new one does not use constraints to infer filters. I could rename the existing one to just `InferFilters` and then combine these two.

Reply:
I thought first we might be able to do so, but it looks okay as it is on second thoughts.